expression/builtin_cast.go

```diff
@@ -1051,32 +1078,33 @@ func (b *builtinCastStringAsIntSig) evalInt(row types.Row) (res int64, isNull bo
 
 	var ures uint64
 	sc := b.ctx.GetSessionVars().StmtCtx
-	if isNegative {
-		res, err = types.StrToInt(sc, val)
-		if err == nil {
-			// If overflow, don't append this warnings
-			sc.AppendWarning(types.ErrCastNegIntAsUnsigned)
-		}
-	} else {
+	if !isNegative {
 		ures, err = types.StrToUint(sc, val)
 		res = int64(ures)
 
 		if err == nil && !mysql.HasUnsignedFlag(b.tp.Flag) && ures > uint64(math.MaxInt64) {
 			sc.AppendWarning(types.ErrCastAsSignedOverflow)
 		}
+	} else if b.inUnionAll && mysql.HasUnsignedFlag(b.tp.Flag) && isNegative {
+		res = 0
+	} else {
+		res, err = types.StrToInt(sc, val)
+		if err == nil {
```

Comment:
`if err != nil`?

Reply:
maybe this old logic is by design https://github.com/pingcap/tidb/pull/3894

Reply:
.....it seems a old bug.
`select cast(-1 as signed)` should not return a warning but master branch does.
I think it should be f2a38a96f46809e13dd6e5f0de4fc124f3e1300d ?